web/src/drafts.ts

```diff
@@ -20,44 +22,86 @@ import * as timerender from "./timerender";
 import * as ui_util from "./ui_util";
 import * as util from "./util";
 
-export function set_count(count) {
+export function set_count(count: number): void {
     const $drafts_li = $(".top_left_drafts");
     ui_util.update_unread_count_in_dom($drafts_li, count);
 }
 
-function getTimestamp() {
+function getTimestamp(): number {
     return Date.now();
 }
 
-export const draft_model = (function () {
-    const exports = {};
+type LocalStorageDraft = {
+    content: string;
+    updatedAt: number;
+} & (
+    | {
+          type: "stream";
+          topic: string;
+          stream_id?: number;
+      }
+    | {
+          type: "private";
+          reply_to: string;
+          private_message_recipient: string;
+      }
+);
+
+// The id is added to the draft in format_drafts in drafts_overlay_ui.
+// We should probably just include it in the draft object itself always?
+type LocalStorageDraftWithId = LocalStorageDraft & {id: string};
+
+const draft_schema = z.intersection(
+    z.object({
+        content: z.string(),
+        updatedAt: z.number(),
+    }),
+    z.union([
+        z.object({
+            type: z.literal("stream"),
+            topic: z.string(),
+            stream_id: z.number().optional(),
+        }),
+        z.object({
+            type: z.literal("private"),
+            reply_to: z.string(),
+            private_message_recipient: z.string(),
+        }),
+    ]),
+);
+
+const drafts_schema = z.record(z.string(), draft_schema);
 
+export const draft_model = (function () {
     // the key that the drafts are stored under.
     const KEY = "drafts";
     const ls = localstorage();
 
-    function get() {
-        return ls.get(KEY) || {};
+    function get(): Record<string, LocalStorageDraft> {
+        const drafts = ls.get(KEY);
+        if (drafts === undefined) {
+            return {};
+        }
+        return drafts_schema.parse(drafts);
```

Comment:
Ideally we parse the drafts one by one so that we can keep valid drafts even if some are invalid. Is there a way to do that with zod? I don't know how to iterate over `drafts` without asserting its type first. (Make all the zod attributes optional and then verify they're defined after?) Would love some advice here.

Reply:
This is a question for #frontend; I don't know that for Zod off the top of my head. I don't think the case where only some drafts are in an invalid format is an important case to handle; this module writes them.

Reply:
You can use [`z.catch`](https://zod.dev/?id=catch), e.g., `z.record(z.string(), draft_schema.catch(undefined)).parse(drafts)`, and then filter out `undefined`.

Reply:
Thanks, that was helpful! Using `catch` now.